File: 238.product-of-array-except-self.py
#
# @lc app=leetcode id=238 lang=python3
#
# [238] Product of Array Except Self
#

# @lc code=start
from typing import *
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = [1]
        right = [1]
        for index,num in enumerate(nums):
            left.append(num*left[index])
        for index,num in enumerate(list(reversed(nums))):
            right.append(right[index]*num)
        return [left[i]*right[len(nums)-i-1] for i in range(len(nums))]
        
# @lc code=end


if __name__ == "__main__":
    print(Solution().productExceptSelf([1,2,3,4]))

File: golang/238.product-of-array-except-self.go
/*
 * @lc app=leetcode.cn id=238 lang=golang
 *
 * [238] Product of Array Except Self
 */
package main

import "fmt"

// @lc code=start
func productExceptSelf(nums []int) []int {
	if len(nums) == 0 {
		return []int{}
	}
	prefixSum := []int{nums[0]}
	for i := 1; i < len(nums); i++ {
		num := nums[i]
		prefixSum = append(prefixSum, prefixSum[i-1]*num)
	}
	suffixSum := make([]int, len(nums))
	suffixSum[len(nums)-1] = nums[len(nums)-1]
	for i := len(nums) - 2; i >= 0; i-- {
		suffixSum[i] = suffixSum[i+1] * nums[i]
	}
	ret := make([]int, len(nums))
	ret[0] = suffixSum[1]
	ret[len(nums)-1] = prefixSum[len(nums)-1-1]

	for i := 1; i < len(nums)-1; i++ {
		ret[i] = prefixSum[i-1] * suffixSum[i+1]
	}
	return ret
}

// @lc code=end

func main() {
	fmt.Println(productExceptSelf([]int{1, 2, 3, 4}))
}
